program-runtime/src/invoke_context.rs

```diff
@@ -643,37 +649,48 @@ impl<'a> InvokeContext<'a> {
         program_indices: &[usize],
         account_indices: &[usize],
         caller_write_privileges: &[bool],
-    ) -> Result<u64, InstructionError> {
+    ) -> ProcessInstructionResult {
         let is_lowest_invocation_level = self.invoke_stack.is_empty();
         if !is_lowest_invocation_level {
             // Verify the calling program hasn't misbehaved
-            self.verify_and_update(instruction, account_indices, caller_write_privileges)?;
+            let result =
+                self.verify_and_update(instruction, account_indices, caller_write_privileges);
+            if result.is_err() {
+                return ProcessInstructionResult {
+                    compute_units_consumed: 0,
+                    result,
+                };
+            }
         }
 
+        let mut compute_units_consumed = 0;
         let result = self
             .push(message, instruction, program_indices, account_indices)
             .and_then(|_| {
                 self.return_data = (*instruction.program_id(&message.account_keys), Vec::new());
                 let pre_remaining_units = self.compute_meter.borrow().get_remaining();
-                self.process_executable_chain(&instruction.data)?;
+                let execution_result = self.process_executable_chain(&instruction.data);
                 let post_remaining_units = self.compute_meter.borrow().get_remaining();
```

Comment:
Looking for a convenient way to write a test that even when `execution_result` is an error here, `post_remaining_units` reflects that compute units were consumed.

Looking at the code here: https://github.com/solana-labs/solana/blob/9c5d82557a965cf23831a3fb2e3b8986d379ce04/programs/bpf_loader/src/lib.rs#L1016-L1022, this should be true as long as `execute_program_jit` updates the `instruction_meter` even on errors, but this would be good to confirm.

Any suggestions for such a test are welcome!

Reply:
Updated: https://github.com/solana-labs/solana/pull/22059/commits/75db0c5af7d4bb33a21fac29becfa519f88cf735